feat(auth): Reuse existing user profile for unknown identity

RyanSkonnord opened the pull request:
When a user supplies unrecognized SSO credentials associated with an email address for an existing user profile, offer them a chance to link their identity to that profile instead of automatically provisioning a new profile.

This is a basis for further feature development. It has function stubs and an incomplete page template that are not reachable unless a feature flag is set. For now, this flag should be set only in a development environment. Added a unit test (`test_unauthenticated_with_existing_user`) which should verify that the stub is unreachable.

RyanSkonnord committed:
feat(auth): Reuse existing user profile for unknown identity

When a user supplies unrecognized SSO credentials associated with an
email address for an existing user profile, offer them a chance to link
their identity to that profile instead of automatically provisioning a
new profile.

Introduce a feature flag. The feature is incomplete and will hit
unimplemented code if the flag is set.

mrbaguvix commented on src/sentry/auth/helper.py:
```diff
@@ -333,6 +334,22 @@ def _get_organization_member(self, auth_identity: AuthIdentity) -> OrganizationM
+                id__in=UserEmail.objects.filter(email__iexact=email).values("user"),
+                is_active=True,
+            ).first()
+        except IndexError:
+            return None
```
Can we add _logger.error_?

RyanSkonnord replied:
That's not necessary. This exception merely indicates that no `User` exists with the identity's email address, which is an expected condition.

mrbaguvix commented on tests/sentry/auth/test_helper.py:
```diff
@@ -42,9 +42,10 @@ def setUp(self):
+        self.email = "test@example.com"
         self.identity = {
             "id": "1234",
-            "email": "test@example.com",
+            "email": self.email,
```
Should we change this?

RyanSkonnord replied:
Can you elaborate? I extracted this into an instance variable to reduce repetition of `self.identity["email"]` below.

JoshFerge commented on tests/sentry/auth/test_helper.py:
```diff
@@ -309,6 +308,25 @@ def test_authenticated(self, mock_render):
 
+    @mock.patch("sentry.auth.helper.render_to_response")
+    @mock.patch("sentry.auth.helper.create_verification_key")
+    def test_unauthenticated_with_existing_user(self, mock_create_key, mock_render):
```
it seems like either we are going to end up having duplicated test behavior between `tests/sentry/web/frontend/test_auth_organization_login.py` and here, or have tested behavior be split across both. i'm not sure what the right resolution is, but I think the integration tests are easier to work with since you don't need so many mocks.

JoshFerge replied:
as discussed offline, probably prudent to enable this flag in the org login tests to confirm existing behavior is preserved when this flag is enabled.

JoshFerge replied:
@RyanSkonnord what was your conclusion on integration tests vs unit tests? It seems like we'd still want integration tests regardless

RyanSkonnord replied:
Added an integration test: https://github.com/getsentry/sentry/pull/28621/commits/833a63361e6518638d3acd78a2561354058dd53b

See also https://github.com/getsentry/sentry/pull/28784 😅

JoshFerge commented on src/sentry/auth/helper.py:
```diff
@@ -457,6 +467,21 @@ def handle_unknown_identity(
+                template = "auth-confirm-link"
+                existing_user = self.user
+            else:
+                existing_user = acting_user or self._get_user(identity)
```
the acting_user will always be None in the case where we'd like to have this new case run. could we simplify by removing this `existing_user` declaration line, and checking if `acting_user` is `None` instead?

JoshFerge replied:
I think there are two other strategies to try and refactor this area: 

1. use a return early pattern
2. in the above where we set the acting_user to None if there's no password, we could instead create a new "op" for the confirm-account scenario (or re-use confirm and check if the user is authed). that would have the advantage of not having to redefine `user`, and removes a bit of complexity here. There might be a bit of duplication from defining the form logic again, but maybe that could be cleaned up too.

RyanSkonnord replied:
Yeah, I wasn't convinced that `acting_user or` did anything but it seemed like a nice defensive step against an unnecessary query.

I'm going to try extracting a helper method out of this code block. Then a return-early pattern will probably clear things up.

JoshFerge commented on src/sentry/auth/helper.py:
```diff
@@ -457,6 +467,21 @@ def handle_unknown_identity(
+                else:
+                    self.request.session.set_test_cookie()
+                    template = "auth-confirm-identity"
+                    existing_user = acting_user
```
this confused me a little bit along with the above, i'm guessing that you wanted to get rid of the `context.update`, but that was more clear to me than redefining a variable.

JoshFerge replied:
maybe it would be most clear if existing_user was first declared as None, and is either set in the first if block to `self._get_user_idenity`, or in this block to `acting_user`

RyanSkonnord replied:
Yes, factoring out the `context.update` was my main motivation. Redefining the `existing_user` variable was sort of opportunistic... it was a natural name for the value being newly queried for, and when I noticed that `"existing_user"` was the existing `context` label it was like, hey, maybe it's actually easier to understand this way?

I think I can make it easier to understand along with the other changes, in the extracted helper method that I mentioned above.

JoshFerge reviewed (commented):
overall looks good 👍🏼 . added some thoughts on structure + test organization

JoshFerge commented on src/sentry/auth/helper.py:
```diff
@@ -457,6 +467,21 @@ def handle_unknown_identity(
+                existing_user = self.user
+            else:
+                existing_user = acting_user or self._get_user(identity)
+                if existing_user and features.has(
```
if the user has a usable password they shouldn't be directed into this block

RyanSkonnord committed:
Extract _dispatch_to_confirmation

RyanSkonnord committed:
Apply automatic migration only for users with no password

RyanSkonnord committed:
Factor acting_user parameter out of _dispatch_to_confirmation

RyanSkonnord committed:
Merge branch 'master' into idp-automatic-migration-base

RyanSkonnord commented:
@JoshFerge I think `_dispatch_to_confirmation` should address most of your stylistic concerns.

(The `"existing_user": existing_user or acting_user` on [line 480](https://github.com/getsentry/sentry/pull/28621/files#diff-e0494c56201878812ba2d58e03a3e6310b0da0b3ed3f5c8ee6c0c9fa243d54aaR480) might be a minor sin, but I like it better than passing `acting_user` to `_dispatch_to_confirmation`. If you hate it, I could revert https://github.com/getsentry/sentry/pull/28621/commits/64f785fb0514bcf376a95ebb839674dab9a52fbf.)

Setting the feature flag on `test_auth_organization_login` does indeed seem to cause some regressions. I'll be working on that next.

JoshFerge commented:
@RyanSkonnord alright sounds good, will wait on your changes to have tests passing with the flag to give it another go-round.

RyanSkonnord commented:
Turns out `test_auth_organization_login` does not have genuine regressions, but the test helper `sentry.testutils.helpers.features.Feature` has some interesting limitations that I've tripped over. I think it may merit an application-wide fix that I'll put in another PR. Will update after I've tinkered with possible fixes a little more.

RyanSkonnord commented:
If we don't want to wait for https://github.com/getsentry/sentry/pull/28686, it should be safe to go ahead. Having manually dummied out the feature flag check, I found no regressions when running `test_auth_organization_login` locally.

JoshFerge reviewed (approved):
having an `acting_user` and `existing_user` is still confusing to me, but I understand wanting these changes to have a small footprint is ideal. maybe we could schedule a team programming session where we do some smaller refactors to this method. see comment about integration tests, but looks good.

RyanSkonnord committed:
Merge branch 'master' into idp-automatic-migration-base

RyanSkonnord committed:
Merge branch 'master' into idp-automatic-migration-base

RyanSkonnord committed:
test: Fix setup in test_flow_authenticated_without_verified_without_password

Change create_user so that calling it with password="" sets up the user
in a state where it does not have a usable password, as the test case
intended. Previously, any falsy value would give it a default password.

This exposes an incorrect assertion. The test case is currently failing.

RyanSkonnord committed:
Fix assertion

RyanSkonnord committed:
Merge branch 'test-auth-organization-login-without-password' into idp-automatic-migration-base

RyanSkonnord committed:
Add new test_auth_organization_login.py case

RyanSkonnord committed:
Merge branch 'master' into idp-automatic-migration-base
